src/browser/history.js

```diff
@@ -0,0 +1,320 @@
+  // Helper function to create cryptographic hash of the content.
+  const sha = string => crypto.subtle.digest("SHA-256",
+                                             stringToBuffer(string)).then(btoa);
+
+  const stub = record => record.constructor.Stub(record);
+  const hash = record => record.constructor.hash(record);
+
+  const write = (record, db, config) => {
+    return db.put(record.toJSON());
+  };
+
+  const read = async(function*(record, db, config) {
+    try {
+      const data = yield db.get(record[PouchDB.id]);
+      return new record.constructor(data);
+    } catch (error) {
+      if (error.status != 404) {
+        throw error
+      }
+      return record;
+    }
+  });
+
+  const edit = async(function*(record, transform, db, config) {
+    const current = yield read(record, db);
+    const next = yield transform(current);
+    return write(next, db, config);
+  });
+
+  const push = x => xs => xs.push(x);
+  const remove = x => xs => xs.remove(x);
+
+  const exclude = x => xs => {
+    const index = xs.indexOf(x)
+    return index < 0 ? xs :
+           index == 0 ? xs.rest() :
+           index + 1 == xs.size ? xs.butLast() :
+           xs.take(index).concat(xs.skip(index + 1))
+  }
+  const include = x => xs => xs.indexOf(x) < 0 ? xs.push(x) : xs
+
+
+  /*
+  ## Interests API
+
+  ### Sites
+
+  Sites store contains records of sites that have being visited by a user.
+  Sites store contains records of the following schema (which could be extended
+  in the future).
+
+  {
+    _id: "site/http://learnyouahaskell.com/introduction#about-this-tutorial",
+    uri: "http://learnyouahaskell.com/introduction#about-this-tutorial",
+    title: "Introduction - Learn You a Haskell for Great Good!",
+    backgroundColor: "rgb(255, 255, 255)",
+    visits: [
+      {
+        start: 1421434329682266,
+        end: 1421434484899,
+        device: "Desktop"
+      }
+    ],
+    tags: [
+      "haskell",
+      "functional"
+    ]
+  }
+  */
+
+  const Visit = Record({
+    type: Type('Visit'),
+    start: TimeStamp,
+    end: Maybe(TimeStamp),
+    id: String, // This is a web view ID in our case.
+    device: String('Desktop'),
+  });
+
+  const Site = Record({
+    [PouchDB.id]: ID,
+    [PouchDB.revision]: Maybe(Revision),
+    type: Type('Site'),
+    uri: URI,
+    title: Maybe(String),
+    visits: List(Visit),
+    tags: List(TagName),
+    icon: Maybe(URI), // Would be better if it was Blob as well.
+    image: Blob,
+  });
+  Site.storeID = 'sites';
+  Site.frequency = ({visits}) => visits.size;
+  Site.from = ({uri, title}) => Site({[PouchDB.id]: `site/${uri}`, uri, title});
+
+  Site.beginVisit = ({time, id, device}) => site =>
+    site.update('visits', push(Visit({start: time, id, device})));
+
+
+  Site.endVisit = ({id, time}) => site => {
+    const index = site.visits.findIndex(visit => visit.id === id);
+    return index < 0 ? site : site.setIn(['visits', index, 'end'], time);
+  };
+
+
+/**
+
+### Quotes
```

Comment:
Can we leave quotes / highlights as is until we actually start using them, I'd like to avoid regressions by renaming stuff. I did manual tests on them and renaming would require doing manual testing again as we otherwise won't spot errors as it's not used yet.